### DISCUSSION TITLE
Possibilities of controlling a Shelly plus add-on

### DISCUSSION THREAD

**Discussion by bartbutenaers** [arendst/Tasmota#18885] [General]
Hi everybody, 

I have to admit that I have never used Tasmota, so perhaps I am going to ask very stupid questions now...

Currently the official Shelly firmware for their [add-on](https://www.shelly.cloud/en-be/products/shop/shelly-plus-add-on) device is very limited.  For example I want to read hc-sr04 ultrasonic distance sensors with it, but that seems not to be possible.  I asked it on their [forum ](https://www.shelly-support.eu/forum/thread/19538-shelly-plus-add-on-to-read-hc-sr04-ultrasonic-proximity-sensor-measurement/?postID=203209) but got not much response unfortunately...

It looks to me that the Shelly add-on devices are [supported ](https://github.com/arendst/Tasmota/issues/7469) by Tasmota, but with my noob Tasmota knowledge I am not sure whether that means it is a fully supported device?  And if it now works out of the box, or perhaps I need to do lots of manual stuff to get it working?

And would it also be possible to read a hc-sr04 sensor with it, when I flash Tasmota on it?  Or is that perhaps completely impossible.    One thing that bothers me is that the add-on device seem to have a single data port, while a hc-sr04 needs both a trigger and an echo pin.  So no idea whether this could ever be possible...

I have also other sensors (like pzem-t004) that I would like to use in the future for energy monitoring.  I assume Shelly will never support stuff like that on their add-on devices, because that would be bad for selling their other energy management devices.  So same question here: would there be a possibility that things like that can ever be possible via Tasmota (if somebody ever finds the free time to implement it), or is that perhaps impossible due to constraints I am not aware of?

Running Tasmota on all my Shelly devices would be a clean uniform setup, that all my less-technical familly members would appreciate ;-)

Thanks for reading!!!!
Bart

**Comment by sfromis** [comment]
Looks like it is not a "controllable device", but an adapter mainly for galvanic isolation. https://github.com/esphome/feature-requests/issues/2034
In principle "invisble" (as in support not being a question), but for specific types of connections, the question may be if the communication can pass through. Maybe that ADC part is a bit special about how signals are to be processed.

The Shelly product page for [Shelly Plus Add-On](https://www.shelly.cloud/en/products/shop/shelly-plus-add-on#node-0h1i0hbsktdrh) does show what looks  like several different sensors, including SR04, implying that the pass-through may be reasonably transparent.

What gpios to use has to match the connector on the Plus devices....
![image](https://github.com/arendst/Tasmota/assets/47082390/a4ad582b-7281-4e6a-8e98-7fa9e1654f27)
Looks like that ESP_DBG_UART pin is gpio 19.

**Reply by bartbutenaers** [reply]
@sfromis,
Thanks for joining here!!

> Looks like it is not a "controllable device"

Yes indeed I should have been more clear.  The add-on is indeed a hat that you need to plug onto another controllable Shelly device (e.g. Shelly plus 1 switch).  There was already a quite in depth analysis for Tasmota [here](https://github.com/arendst/Tasmota/issues/7469), but I don't really understand it quite well to be honest...

> The Shelly product page for [Shelly Plus Add-On](https://www.shelly.cloud/en/products/shop/shelly-plus-add-on#node-0h1i0hbsktdrh) does show what looks like several different sensors, including SR04

Do you mean this sensor on their product page:

![image](https://github.com/arendst/Tasmota/assets/14224149/19e2ef7f-ca14-4cd5-863d-e415b91e9f02)

Because that is no hc-sr04, but an [URM09](https://wiki.dfrobot.com/URM09_Ultrasonic_Sensor_Gravity_Trig_SKU_SEN0388) sensor, which has analog value as output.  They can support that sensor, because you can use their add-on to read the analog value with their adc converter.

I bought a box of add-on devices, because the product page mentioned that it supports "ultrasonic proximity sensors".  But when they arrived I realized that I could only use URM09 sensors.  And not my hc-sr04 sensors (or my waterproof [jsn-sr04t](https://www.tinytronics.nl/shop/en/sensors/distance/waterproof-ultrasonic-sensor-jsn-sr04t) sensors) ;-(

> Looks like that ESP_DBG_UART pin is gpio 19.

That is unfortunately too much out of my comfort zone to understand if this is good or bad news for me.  So I would appreciate if you could explain this a bit more in simple terms...

**Reply by sfromis** [reply]
The other Tasmota discussion you link to is for the older non-plus variant of the addon, but looks pretty similar, except that this newer one has more pins.

While the sensor in the image is not SR04 then, I'd still *guess* that it would be possible to support SR04 using two of the inputs, as long as you work out what inputs are connected to what pins. Of course, I cannot be sure without having tried it.

The bit about gpio 19 may not be important, it was just to explain the 7-pin connector. It may not even be relevant to the addon adapter, but that depends on what the inputs are connected to. But since it appears to have 5 inputs, and only 3 or 4 usable connections to the "mothership", it may be a question of not all working at the same time, and *potentially* having one of them connected to gpio 19. The point is that the connector has gpio 0, two serial pins and that "debug" pin, and I do not know what is connected to what, but likely this may be easy to work out experimentally.

**Reply by barbudor** [reply]
Not sure if that applies to that model of Shelly but I seem to recall that gpio available on the add-on connector are not directly gpio and may not be reversible.
If I'm correct, the DS18X20 needs actually 2 gpio on Shelly to be able to write and read the one wire bus.

So if the SR04 works in a similar way, sending the trigger on a GPIO, the reversing the GPIO to input to read the pulse, this may not work on the Shelly.

But that was also for non-Plus models
Not sure how the add-on support is different on the Plus

**Reply by bartbutenaers** [reply]
Thanks all for the replies!!!
So from the comments I conclude that there it might not work, even if somebody with knowledge about this would have time to investigate this.  That is a pitty because I got no response from Shelly anymore meanwhile ;-(

**Reply by PhaseAngle** [reply]
DS18X20 only needs 1 GPIO, VCC and GND.
The 3 DS18X20 inputs are connected together and why you can connect up to 5 DS18X20 sensors with Shelly firmware.
[https://kb.shelly.cloud/knowledge-base/shelly-plus-add-on#Deviceelectricalinterfaces-1-Wireinterface](url)
This would also account for the single DHT22 in place of the DS18X20 bus

![image](https://github.com/arendst/Tasmota/assets/42715477/c6466877-6f5b-4232-a85e-c7e8e7851923)
 Then there is just 1 analogue in and 1 digital in.

**Comment by pgorod** [comment]
@sfromis @barbudor if it would help, as a contribution to this community, I would be willing to buy one of you a new Shelly Plus Add-on and send it to your address, so you could tinker with it.

You would only owe me an honest try at getting it working. Then give me a working template ( or your honest assessment "sorry, it would be too much work to get this working") and you get to keep the add-on. I am assuming you have a Shelly Plus 1PM to try it on, do you?

If one of you wants to accept the offer (and challenge) we can figure out a private way for you to tell me which address I can send it to. Thanks

**Comment by pgorod** [comment]
@ascillato2 @blakadder @arendst would any of you be interested in my offer above? Thanks

**Reply by blakadder** [reply]
no

**Comment by lordjaxom** [comment]
Just in case someone find this discussion (like me) while searching for Shelly Plus 1 Addon support in Tasmota to get a DHT22 running with it. I managed to get it working using the following Pin configuration:

GPIO0 DHT11_o
GPIO1 AM2301

Full Template: {"NAME":"Shelly Plus 1 ","GPIO":[1280,1216,0,0,192,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,32,224,0,0,0,0,0,4736,4705,0,0,0,0,0,0],"FLAG":0,"BASE":1}

**Reply by stehof2** [reply]
I’ve been searching and trying for hours now, but your post, almost 1 year later saved my day, thanks a lot :)

**Comment by 13** [comment]
Here the Shelly Plus Addon pins:

DATA = GPIO1
ANALOG IN = GPIO3
DIGITAL IN = GPIO19

**Reply by completelyClueless** [reply]
sorry - what do I have to select when in screen "configure template" when using multiple DS18B20 ?   I2C SDA/SDL ?

**Reply by sfromis** [reply]
No, not I2C pins, but select "DS18B20". Try with one first, to see if it works through the add-on. Tasmota supports multiple DS18B20 on the same wire.

**Reply by completelyClueless** [reply]
hi, 
thx for your quick response;
tried that already - not working :-(
stepped back to 11.1.0 (to have DS18B20_o available) => does not work

**Reply by barbudor** [reply]
Shelly have some specific design that makes the GPIO as non reversible (which is required for the 1wire protocol)
I don't know where this is documented (and if it is) but if I'm not wrong you need 2 gpios (in your case 1 and 19), connect them together and to the data pin of the DS, one should be set as DS18X20_o and the other one  as DS18X20 normal
I would expect the _o to be on gpio 19 but if not working, try the other way round

Edit fix typo: gpio19

**Reply by sfromis** [reply]
> I would expect the _o to be on gpio

The `_o` pin is only available on ESP8266, and the plus add-on is for ESP32 devices. Since it was removed from ESP32, that implies that it was not seen to be useful.

A comment from an earlier source version says `// ESP32 don't support dual pin mode`

**Reply by barbudor** [reply]
Not sure that's the right way to interpret that
Not supported is not equivalent to not needed

Can you, or anyone, confirm having a DS18X20 working on Shelly ESP32 based device?

**Reply by sfromis** [reply]
I was not trying to imply "not needed", only that it (dual pin mode for DS18x20) is not available in current Tasmota for ESP32.

And no, I do not have the add-on.

**Reply by completelyClueless** [reply]
I can support with testing - I have a Shelly 1 Plus connected with the Shelly Addon

**Comment by HB787** [comment]
I want to connect two DS18B20 temp sensors on the Shelly plus add-on, in combination with a Shelly plus 2PM. The DHT22 is working using the pins from @lordjaxom. I also tried the SR501 PIR without success. In the Shelly firmware it was all working.
Is there any solution on the table for this, because the add-on device is popular.

**Reply by 13** [reply]
Shelly Plus 2PM with Shelly Plus Addon features a different pinout:

**DIGITAL IN** is mapped to **GPIO22**, but I couldn't locate the analog input pin.

Has anyone had success identifying or using the analog input on this device?

**Comment by joeH68** [comment]
Hi there, thinking of using tasmotized Shelly Plus 2PM + Shelly Plus Add On to control awning with wind (analog 0-10 VDC input) and O/C (magnetic) sensors. I'm experienced with other Shellies but not these ones. 

1. is there a "standard" template I may use as base?  what change(s) would be required?
2. with regards to rules, what event triggers should I look for to react to these sensors?

Any hints will be much appreciated

**Reply by sfromis** [reply]
There can be no "standard template", as template config of pins depends on what they are to be used for, what is connected where to the add-on. I'd not expect it to support 0-10V without the voltage being divided down to supported levels, read the specs. Typically, magnetic "sensors" are simple on/off which can be configured with a suitable Tasmota config for a Switch or Button input, depending on the exact behavior. No different from having a pushbutton connected.

**Comment by joeH68** [comment]
thanks @sfromis. Order is on its way, will report findings. 

Not clear about "I'd not expect it to support 0-10V without the voltage being divided down to supported levels".
The [Plus Addon specs](https://www.shelly-support.eu/filebase/entry-download/230-shelly-plus-addon-bedienungsanleitung-mehrsprachig/) states "Analog input range: 0 - 10 VDC". Do you mean a different document?

**Reply by sfromis** [reply]
Well, I explicitly had "read the specs" after what you quoted, meaning not just assuming that voltages **much** higher than the ESP chip supports will work. And yes, whatever is in the specs should be ok.

**Comment by Ruhz69** [comment]
Dear all, I would be interested as well, I have a Shelly plus 1 and an add on with 3 DS18B20. If you get me a list of what to try, I would do so.

**Comment by Kolia56** [comment]
I'm trying as well to have DS18x20 connected to a Shelly Plus 1 device with addon. No success so far. I setup GPIO pin 1 to DS18x20 as described here [https://github.com/arendst/Tasmota/discussions/18885#discussioncomment-7295482](url). No luck. To what GPIO pin should it be connected in order to get DS18x20 recognized? Any other settings I should be aware off. Is it supposed to be supported?
Thank you

**Comment by Kolia56** [comment]
Please see this new [discussion thread](https://github.com/arendst/Tasmota/discussions/20580). One unit is on it's way to @arendst

**Comment by Kolia56** [comment]
As of commit 4bc7b4e DS18x20 are now supported.
